reasoner/planner/AnswerCountEstimator.java

```diff
@@ -447,37 +456,39 @@ public boolean isRelation() {
             }
 
             @Override
-            long estimateAnswers(Set<Variable> variableFilter) {
-                long singleRelationEstimate = 1L;
+            double estimateAnswers(Set<Variable> variableFilter) {
+                double singleRelationEstimate = 1L;
                 Map<Label, Integer> queriedRolePlayerCounts = new HashMap<>();
+                double typeBasedUpperBound = 1L;
                 for (Variable v : variableFilter) {
                     if (v.isThing() && rolePlayerTypes.containsKey(v.asThing())) {
-                        Label vType = this.rolePlayerTypes.get(v);
-                        queriedRolePlayerCounts.put(vType, 1 + queriedRolePlayerCounts.getOrDefault(vType, 0));
+                        Label role = this.rolePlayerTypes.get(v);
+                        queriedRolePlayerCounts.put(role, 1 + queriedRolePlayerCounts.getOrDefault(role, 0));
+                        typeBasedUpperBound *= typeMaximums.get(v);
```

Comment:
Fix: Was doing type based estimates on an assumption that variables playing the same role would be interchangeable. Now it considers each variable separately.